testcases/crypto/ec_func.c

```diff
@@ -162,6 +162,566 @@ _signVerifyParam signVerifyInput[] = {
+    case 5: return CURVE384_LENGTH/8;
+    case 6: return CURVE512_LENGTH/8;
+    case 7: return CURVE512_LENGTH/8;
+    case 8: return CURVE192_LENGTH/8;
+    case 9: return CURVE224_LENGTH/8;
+    case 10: return CURVE256_LENGTH/8;
+    case 11: return CURVE384_LENGTH/8;
+    case 12: return CURVE521_LENGTH/8+1;
+    }
+
+    return 0;
+}
+
+static CK_RV curve_supported(const unsigned char *name)
+{
+    if (name[strlen(name)-2] == 'r' || name[strlen(name)-2] == 'v')
+        return 1;
+    else
+        return 0;
+}
+
+/**
+ * A test is skipped for the ep11token, when the derived key length
+ * shall be bigger than the shared secret (z-value). The ep11token is
+ * currently based on PKCS#11 v2.20 where no KDF is considered.
+ * This restriction comes from the ep11 host library.
+ */
+static unsigned int too_many_key_bytes_requested_ep11(unsigned int curve, unsigned int kdf,
+        unsigned int keylen)
+{
+    if (!is_ep11_token(SLOT_ID))
+        return 0;
+
+    if (keylen <= curve_len(curve))
+        return 0;
+    else
+        return 1;
+}
+
+/**
+ * A test is skipped, when no KDF is used and the derived key length
+ * shall be bigger than the shared secret (z-value). Without a KDF, max
+ * z-length key bytes can be derived.
+ */
+static unsigned int too_many_key_bytes_requested(unsigned int curve, unsigned int kdf,
+        unsigned int keylen)
+{
+    if (kdf > 0 || keylen <= curve_len(curve))
+        return 0;
+    else
+        return 1;
+}
+
+/*
+ * Generate EC key-pairs for parties A and B.
+ * Derive shared secrets based on Diffie Hellman key agreement defined in PKCS#3.
+ */
+CK_RV run_DeriveECDHKey()
+{
+    CK_SESSION_HANDLE session;
+    CK_MECHANISM mech;
+    CK_FLAGS flags;
+    CK_OBJECT_HANDLE publ_keyA = CK_INVALID_HANDLE, priv_keyA = CK_INVALID_HANDLE;
+    CK_OBJECT_HANDLE publ_keyB = CK_INVALID_HANDLE, priv_keyB = CK_INVALID_HANDLE;
+    CK_OBJECT_HANDLE secret_keyA = CK_INVALID_HANDLE, secret_keyB = CK_INVALID_HANDLE;
+    CK_BYTE user_pin[PKCS11_MAX_PIN_LEN];
+    CK_ULONG user_pin_len;
+    CK_RV rc = CKR_OK;
+    CK_ECDH1_DERIVE_PARAMS ecdh_parmA, ecdh_parmB;
+    CK_BBOOL true = CK_TRUE;
+    CK_BYTE pubkeyA_value[256];
+    CK_BYTE pubkeyB_value[256];
+    CK_BYTE secretA_value[80000]; // enough space for lengths in secret_key_len[]
+    CK_BYTE deriveB_value[80000];
+    CK_OBJECT_CLASS class = CKO_SECRET_KEY;
+    CK_KEY_TYPE key_type = CKK_GENERIC_SECRET;
+
+    testcase_begin("starting run_DeriveECDHKey...");
+    testcase_rw_session();
+    testcase_user_login();
+
+    if (!mech_supported(SLOT_ID, CKM_EC_KEY_PAIR_GEN)) {
+        testcase_skip("Slot %u doesn't support CKM_EC_KEY_PAIR_GEN\n", (unsigned int) SLOT_ID);
+        goto testcase_cleanup;
+    }
+    if (!mech_supported(SLOT_ID, CKM_ECDH1_DERIVE)) {
+        testcase_skip("Slot %u doesn't support CKM_ECDH1_DERIVE\n", (unsigned int) SLOT_ID);
+        goto testcase_cleanup;
+    }
+
+    for (int i=0; i<NUMEC; i++) {
```

Comment:
Do not declare int i inside the for, please do it outside. This is causing Travis CI build to fail.